core/lib/server/pub.ts

```diff
@@ -589,3 +600,574 @@ export const deletePub = async (pubId: PubsId) =>
+									sql<number>`pub_tree.depth + 1`.as("depth"),
+									// this is a standard way to detect cycles
+									// we keep track of the path we've taken so far,
+									// and set a flag if we see a pub that is already in the path
+									// https://www.postgresql.org/docs/current/queries-with.html#QUERIES-WITH-CYCLE
+									sql<boolean>`pubs.id = any(pub_tree."path")`.as("isCycle"),
+									sql<PubsId[]>`array_append(pub_tree."path", pubs.id)`.as(
+										"path"
+									),
+								])
+								.where("pub_tree.depth", "<", depth)
+								.where("pub_tree.isCycle", "=", false)
+								.$if(cycle === "exclude", (qb) =>
+									// this makes sure we don't include the first pub that is part of a cycle
+									qb.where(sql<boolean>`pubs.id = any(pub_tree.path)`, "=", false)
+								)
+						)
+					)
+			)
+			// this CTE finds the top level pubs and limits the result
+			// counter intuitively, this is CTE is referenced in the above `withRecursive` call, despite
+			// appearing after it. This is allowed in Postgres. See https://www.postgresql.org/docs/current/sql-select.html#SQL-WITH
+			// this is mostly because `kysely` does not allow you to put a normal CTE before a recursive CTE
+			.with("root_pubs_limited", (cte) =>
+				cte
+					.selectFrom("pubs")
+					.selectAll("pubs")
+					.$if(Boolean(props.pubId), (qb) => qb.where("pubs.id", "=", props.pubId!))
+					.$if(Boolean(props.stageId), (qb) =>
+						qb
+							.innerJoin("PubsInStages", "pubs.id", "PubsInStages.pubId")
+							.where("PubsInStages.stageId", "=", props.stageId!)
+					)
+					.$if(Boolean(props.communityId), (qb) =>
+						qb.where("pubs.communityId", "=", props.communityId!)
+					)
+					.$if(Boolean(props.pubTypeId), (qb) =>
+						qb.where("pubs.pubTypeId", "=", props.pubTypeId!)
+					)
+					.$if(Boolean(limit), (qb) => qb.limit(limit!))
+					.$if(Boolean(offset), (qb) => qb.offset(offset!))
+			)
+			.selectFrom("pub_tree")
+			.select((eb) => [
+				"pub_tree.pubId",
+				"pub_tree.parentId",
+				"pub_tree.pubTypeId",
+				"pub_tree.depth",
+				"pub_tree.stageId",
+				"pub_tree.communityId",
+				"pub_tree.isCycle",
+				"pub_tree.path",
+				"pub_tree.createdAt",
+				// we return the updatedAt of the latest value, because the updatedAt of the pub itself
+				// does not really change over time
+				eb.fn
+					.coalesce(eb.fn.max("pub_tree.valueUpdatedAt"), "pub_tree.updatedAt")
+					.as("updatedAt"),
+				jsonArrayFrom(
+					eb
+						.selectFrom("pub_tree as inner")
+						.select((eb) => [
+							"inner.valueId as id",
+							"inner.fieldId",
+							"inner.value",
+							"inner.relatedPubId",
+							"inner.valueCreatedAt as createdAt",
+							"inner.valueUpdatedAt as updatedAt",
+							jsonObjectFrom(
+								eb
+									.selectFrom("pub_tree as inner2")
+									.select(["inner2.schemaName", "inner2.slug"])
+									.whereRef("inner2.valueId", "=", "inner.valueId")
+									.limit(1)
+							).as("field"),
```

Comment:
I vote for the flat version, not nested in `field`. would also be good for the upcoming `isTitle` field, I think it just reads a little more naturally being attributes of the value